core/src/main/java/cz/o2/proxima/time/WatermarkEstimator.java

```diff
@@ -15,178 +15,29 @@
-        TimestampSupplier timestampSupplier) {
-
-      this.durationMs = durationMs;
-      this.stepMs = stepMs;
-      this.allowedTimestampSkew = allowedTimestampSkew;
-      this.minWatermark = minWatermark;
-      this.timestampSupplier = timestampSupplier;
-    }
-
-    public Builder withDurationMs(long durationMs) {
-      return new Builder(durationMs, stepMs, allowedTimestampSkew, minWatermark, timestampSupplier);
-    }
-
-    public Builder withStepMs(long stepMs) {
-      return new Builder(durationMs, stepMs, allowedTimestampSkew, minWatermark, timestampSupplier);
-    }
-
-    public Builder withAllowedTimestampSkew(long allowedTimestampSkew) {
-      return new Builder(durationMs, stepMs, allowedTimestampSkew, minWatermark, timestampSupplier);
-    }
-
-    public Builder withMinWatermark(long minWatermark) {
-      return new Builder(durationMs, stepMs, allowedTimestampSkew, minWatermark, timestampSupplier);
-    }
-
-    public Builder withTimestampSupplier(TimestampSupplier timestampSupplier) {
-      return new Builder(durationMs, stepMs, allowedTimestampSkew, minWatermark, timestampSupplier);
-    }
-
-    public WatermarkEstimator build() {
-      return new WatermarkEstimator(
-          durationMs, stepMs, allowedTimestampSkew, minWatermark, timestampSupplier);
-    }
-  }
-
-  public static Builder newBuilder() {
-    return new Builder();
-  }
-
-  private final long stepMs;
-  private final TimestampSupplier timestampSupplier;
-  private final long[] stepDiffs;
-  private final long allowedTimestampSkew;
-  private final AtomicLong lastRotate;
-  private final AtomicInteger rotatesToInitialize;
-  private final AtomicLong watermark;
-  private final AtomicLong lastStatLogged = new AtomicLong();
-
-  @VisibleForTesting
-  WatermarkEstimator(
-      long durationMs,
-      long stepMs,
-      long allowedTimestampSkew,
-      long minWatermark,
-      TimestampSupplier supplier) {
-
-    this.stepMs = stepMs;
-    this.allowedTimestampSkew = allowedTimestampSkew;
-    this.timestampSupplier = Objects.requireNonNull(supplier);
-    this.watermark = new AtomicLong(minWatermark);
-    Preconditions.checkArgument(durationMs > 0, "durationMs must be positive");
-    Preconditions.checkArgument(stepMs > 0, "stepMs must be positive");
-    Preconditions.checkArgument(
-        durationMs / stepMs * stepMs == durationMs, "durationMs must be divisible by stepMs");
-    stepDiffs = new long[(int) (durationMs / stepMs) + 1];
-    for (int i = 0; i < stepDiffs.length; i++) {
-      stepDiffs[i] = 0;
-    }
-    rotatesToInitialize = new AtomicInteger(stepDiffs.length - 1);
-    lastRotate = new AtomicLong(supplier.get() - stepMs);
-  }
+/** Estimates watermark according to incoming stream elements. */
+public interface WatermarkEstimator extends WatermarkSupplier, Serializable {
 
   /**
-   * Accumulate given timestamp at current processing time.
+   * Returns monotonically increasing estimated watermark.
```

Comment:
I think the watermark is not strictly required to be monotonically increasing. The watermark can decrease in case when the consumption of commit log is restarted (e.g. from committed position). That might be done by creating new estimator, do we have this covered with tests?

Reply:
Not sure what you think. I meant that one estimator instance should always return mon. increasing watermark. If source is restarted then a new instance of estimator is created and estimator starts from older watermark, but again returns mon. increasing stamps.

Reply:
Source is restarted only on program restart isn't it?

Reply:
Not necessarily. Under beam model yes, but replication controller can do that at any time. Proxima's model is sonewhat generalization of the beam model in this sense.

Reply:
@dmvk any thoughts?

Reply:
We have to make sure to reinstantiate the estimator after rebalances.